### ISSUE TITLE
[clang] False-positive warning for integer promotion [-Wimplicit-int-conversion]

### ISSUE THREAD

**Issue by Mephistophiles** [llvm/llvm-project#149288]
```c
#include <stdint.h>

_Static_assert(__builtin_types_compatible_p(typeof(UINT8_C(1)), int), "");
_Static_assert(__builtin_types_compatible_p(typeof(UINT8_C(1)), uint8_t) == 0, "");

int foo(uint8_t b)
{
    uint8_t a = ((uint8_t)1) << b;
    return __builtin_types_compatible_p(typeof(((uint8_t)1) << b), int);
}

```
https://godbolt.org/z/s1f64449b

Code snippet gives a warning:
```
<source>:8:30: warning: implicit conversion loses integer precision: 'int' to 'uint8_t' (aka 'unsigned char') [-Wimplicit-int-conversion]
    8 |     uint8_t a = ((uint8_t)1) << b;
      |             ~   ~~~~~~~~~~~~~^~~~
```

However, this behavior is not considered an error in gcc (there is no warning):
https://godbolt.org/z/d1KjrbKdd

**Comment by frederick-vs-ja** [comment]
I think Clang is doing the right thing.

**Comment by shafik** [comment]
Shift will perform promotions on its operands see [expr.shift p1](https://eel.is/c++draft/expr.shift#1):

> ... and integral promotions are performed. ... result is that of the promoted left operand

and based on [conv.prom](https://eel.is/c++draft/conv.prom) the result will be `int`.
